### PR TITLE
updated description of input features

### PR DESCRIPTION
**Pull request by sanjaydasgupta**

### CODE REVIEW

File: docs/configuration/large_language_model.md [hunk 1]
```diff
@@ -88,21 +88,59 @@ ludwig train ...
...
+Because of the way LLMs work, their input is always a single chunk of `text`. But this input can be 
+built up from one or more columns of the input dataset and optional interspersed static text. 
+The following two examples illustrate the range of possibilities.
 
-If no `prompt` template is provided, this feature must correspond to a column
-in the input dataset. If a prompt template is provided, the rendered prompt
-will be used as the input feature value during training and inference.
+### Single Dataset Column
+
+If input to the LLM is just the content of a single dataset column (without any other prefixed or
```
Comment (alexsherstinsky) [inline docs/configuration/large_language_model.md:97]: @sanjaydasgupta I would rephrase this as "If it is intended for the input to the LLM to be just the" [and the rest is OK]

File: docs/configuration/large_language_model.md [hunk 2]
```diff
@@ -88,21 +88,59 @@ ludwig train ...
...
+suffixed text), no `prompt` template should be provided and the `name` of the feature must 
+correspond to a column in the input dataset. See the following example.
```
Comment (alexsherstinsky) [inline docs/configuration/large_language_model.md:99]: @sanjaydasgupta likely, I would say here "and the `name` of the feature must correspond to that specific database column".

File: docs/configuration/large_language_model.md [hunk 3]
```diff
@@ -88,21 +88,59 @@ ludwig train ...
...
 configuration options.
 
+### Multiple Dataset Columns with Interspersed Static Text
+
+If input to the LLM must include content from one or more dataset columns and static text, 
+then a `prompt` template should be provided to specify how the content of the dataset's 
+column(s) should be formatted for the LLM. See the following example.
+
+```yaml
+input_features:
```
Comment (alexsherstinsky) [inline docs/configuration/large_language_model.md:120]: @sanjaydasgupta I believe that this is not the case.  What happens is that when the prompt template is provided, this is only one input feature, and its name could be anything -- e.g., "my_input" (any chance you could please verify this?).  Thank you.

**Review by alexsherstinsky — CHANGES_REQUESTED** [review]
@sanjaydasgupta Thank you for doing this.  I think your direction is correct in general (I just left a few clarifying suggestions).  Thank you.

**Comment by sanjaydasgupta** [comment]
Thanks @alexsherstinsky. I will work with your suggestions during the weekend.

File: docs/configuration/large_language_model.md (hunk 3 above)
Comment (sanjaydasgupta) [reply docs/configuration/large_language_model.md]: Ah! You just corrected something I had misunderstood before.

**Comment by sanjaydasgupta** [comment]
Hi @alexsherstinsky, please let me know if you think the new description is too wordy.

File: docs/configuration/large_language_model.md [hunk 4]
```diff
@@ -88,21 +88,66 @@ ludwig train ...
...
+    [/INST]
+
+input_features:
+  - name: prompt
+    type: text
+```
+
+There are a couple of things to note here:
+- the prompt `template` contains named placeholders (`context` and `question`) for content 
+from the dataset's columns
```
Comment (alexsherstinsky) [inline docs/configuration/large_language_model.md:141]: @sanjaydasgupta Just to note, these values can also come from a dictionary.  For example, we can use this prompt template and `format()` it with those dictionary name-value pairs.  Thanks!

File: docs/configuration/large_language_model.md [hunk 5]
```diff
@@ -88,21 +88,66 @@ ludwig train ...
 
 ## Input Features
 
-Currently, the LLM model type only supports a single input feature of type `text`.
+Because of the way LLMs work, their input is always a single chunk of `text`. But this input can be
```
Comment (alexsherstinsky) [inline docs/configuration/large_language_model.md:91]: @sanjaydasgupta Please consider the following changes:
1) "their input is always a single chunk of `text`." -> "they accept exactly one input, which is some amount of `text`."
2) Remove "But": "This input can be [...etc...]".
3) Change "is immaterial, and is not" to "is immaterial; it is not".

Everything else looks great!  Thank you so much!

Comment (sanjaydasgupta) [reply docs/configuration/large_language_model.md]: Done!

File: docs/configuration/large_language_model.md (hunk 4 above)
Comment (sanjaydasgupta) [reply docs/configuration/large_language_model.md:141]: I wasn't sure if you wanted any change here. Please clarify.

File: docs/configuration/large_language_model.md (hunk 2 above)
Comment (sanjaydasgupta) [reply docs/configuration/large_language_model.md]: I took the liberty of changing your `database` to `dataset` for the sake of consistency throughout the document

File: docs/configuration/large_language_model.md (hunk 1 above)
Comment (sanjaydasgupta) [reply docs/configuration/large_language_model.md]: Done

**Comment by sanjaydasgupta** [comment]
Hi @alexsherstinsky, can you please review [this](https://github.com/ludwig-ai/ludwig-docs/pull/356#discussion_r1545519865) and [this](https://github.com/ludwig-ai/ludwig-docs/pull/356#discussion_r1545252270).

**Review by alexsherstinsky — APPROVED** [review]
@sanjaydasgupta This LGTM!  Thank you very much for making these clarifications in the documentation.  I would like to please get a glance-over by @arnavgarg1 before merging.  Thank you very much!

File: docs/configuration/large_language_model.md [hunk 6]
```diff
@@ -88,21 +88,66 @@ ludwig train ...
 
 ## Input Features
 
-Currently, the LLM model type only supports a single input feature of type `text`.
+Because of the way LLMs work, they accept exactly one input, which is some amount of `text`.
```
Comment (arnavgarg1) [inline docs/configuration/large_language_model.md:91]: nit: `of type text` instead of `some amount of text`? I think it also gets the point across of it needing to be a text type feature in a Ludwig config

File: docs/configuration/large_language_model.md [hunk 7]
```diff
@@ -88,21 +88,66 @@ ludwig train ...
...
+interspersed static text. The following two examples illustrate the range of possibilities.
 
-If no `prompt` template is provided, this feature must correspond to a column
-in the input dataset. If a prompt template is provided, the rendered prompt
-will be used as the input feature value during training and inference.
+### Single Dataset Column
+
+If it is intended for the input to the LLM to be just the content of a single dataset column (without 
+any other prefixed or suffixed text), no `prompt` template should be provided and the `name` of the 
+feature must correspond to that specific dataset column. See the following example.
```
Comment (arnavgarg1) [inline docs/configuration/large_language_model.md:99]: Maybe we can also mention that it is still possible to wrap the single column with a static prompt template if desired.

File: docs/configuration/large_language_model.md [hunk 8]
```diff
@@ -88,21 +88,66 @@ ludwig train ...
...
+    Answer the question using only the provided context.
+    <</SYS>>
+    
+    ### Context:
+    {context}
+    
+    ### Question:
+    {question}
+    [/INST]
+
```
Comment (arnavgarg1) [inline docs/configuration/large_language_model.md:133]: Perhaps add the end of the prompt template, you may also want to add:

```
### Response:
```

The reason is to illustrate the LLMs are auto regressive models and will continue to produce tokens from the previous token. It generally tends to perform better when templates are structured this way!

File: docs/configuration/large_language_model.md [hunk 9]
```diff
@@ -88,21 +88,66 @@ ludwig train ...
...
+input_features:
+  - name: prompt
+    type: text
+```
+
+There are a couple of things to note here:
```
Comment (arnavgarg1) [inline docs/configuration/large_language_model.md:139]: nit: `note in this example`?

File: docs/configuration/large_language_model.md [hunk 10]
```diff
@@ -88,21 +88,66 @@ ludwig train ...
...
+- the prompt `template` contains named placeholders (`context` and `question`) for content 
+from the dataset's columns 
+- the `name` of the `input_feature` (`prompt` here) is immaterial; it is not the name 
+of a dataset column as in the previous example. This example uses `prompt` just to 
+emphasize that the formatted output obtained by applying the prompt template to the 
+dataset columns is the input.
+
+Also note that a prompt template can be used even with a single dataset column. 
+But that is necessary only when static text is required to be prefixed and/or 
+suffixed to the content of the dataset column.
```
Comment (arnavgarg1) [inline docs/configuration/large_language_model.md:149]: Maybe we can move this up into the single column dataset section?

File: docs/configuration/large_language_model.md [hunk 11]
```diff
@@ -88,21 +88,66 @@ ludwig train ...
...
+    type: text
+```
+
+There are a couple of things to note here:
+- the prompt `template` contains named placeholders (`context` and `question`) for content 
+from the dataset's columns 
+- the `name` of the `input_feature` (`prompt` here) is immaterial; it is not the name 
+of a dataset column as in the previous example. This example uses `prompt` just to 
+emphasize that the formatted output obtained by applying the prompt template to the 
+dataset columns is the input.
```
Comment (arnavgarg1) [inline docs/configuration/large_language_model.md:145]: While true, I think this may be confusing to readers/first time users. What if perhaps we actually just use `question` as the column name but then say it is a placeholder and will be substituted by the value from the prompt template? This just reduces scope for unintended user errors!

**Review by arnavgarg1 — COMMENTED** [review]
Thanks for making this part of the docs clearer! I just left a few minor comments for consideration

File: docs/configuration/large_language_model.md (hunk 6 above)
Comment (sanjaydasgupta) [reply docs/configuration/large_language_model.md:91]: Done

File: docs/configuration/large_language_model.md (hunk 8 above)
Comment (sanjaydasgupta) [reply docs/configuration/large_language_model.md:133]: Done

**Comment by sanjaydasgupta** [comment]
Hi @arnavgarg1, thank you very much for your very insightful comments. I believe documentation should be widely reviewed, and having another pair of eyes was truly helpful.

Since this was a particularly troublesome section of the documentation, I though (after reading all of your comments) that it would be best to add one more yaml example. So I have added one more case (one dataset column + template). That enabled all of your structural comments to be handled in a much better way, and should give readers a much better understanding of the possible variations.

I request you and @alexsherstinsky to please review my changes critically once more and let me know if anything has fallen between the cracks.

File: docs/configuration/large_language_model.md (hunk 11 above)
Comment (sanjaydasgupta) [reply docs/configuration/large_language_model.md:145]: After making the extensive changes referred in my [last comment below](https://github.com/ludwig-ai/ludwig-docs/pull/356#issuecomment-2030970301) I felt that `prompt` was sounding ok. However, if you still think `question` will be better, please let me know and I will change it.

File: docs/configuration/large_language_model.md (hunk 10 above)
Comment (sanjaydasgupta) [reply docs/configuration/large_language_model.md:149]: Please see my [last comment below](https://github.com/ludwig-ai/ludwig-docs/pull/356#issuecomment-2030970301)

File: docs/configuration/large_language_model.md (hunk 9 above)
Comment (sanjaydasgupta) [reply docs/configuration/large_language_model.md:139]: Please see my [last comment below](https://github.com/ludwig-ai/ludwig-docs/pull/356#issuecomment-2030970301)

File: docs/configuration/large_language_model.md (hunk 7 above)
Comment (sanjaydasgupta) [reply docs/configuration/large_language_model.md:99]: Please see my [last comment below](https://github.com/ludwig-ai/ludwig-docs/pull/356#issuecomment-2030970301)

File: docs/configuration/large_language_model.md [hunk 12]
```diff
@@ -88,20 +88,83 @@ ludwig train ...
...
+If the input to the LLM must be created by prefixing and/or suffixing some static text 
+to the content of one dataset column, then a `prompt` `template` should be provided to specify how 
+the content of the chosen column should be formatted for the LLM. See the following example.
+
+```yaml
+prompt:
+  template: |
+    Translate into French: 
+    {english_input}
+
```
Comment (arnavgarg1) [inline docs/configuration/large_language_model.md:123]: Love it! Can we update it to add this?

```
prompt:
  template: |
    Translate into French

    Input: {english_input}
    Translation: 
```

**Review by arnavgarg1 — APPROVED** [review]
I love the clarity and the extra example YAML helps bring it home. Thanks for taking on one of the hardest parts of our documentation and simplifying it! I left one final comment, but approving since this is ready to go after that minor change. 

Thanks for your contribution @sanjaydasgupta

**Comment by sanjaydasgupta** [comment]
@arnavgarg1 @alexsherstinsky changes completed

**Comment by arnavgarg1** [comment]
Thanks! Will defer to Alex to do a final pass before merging

**Review by alexsherstinsky — APPROVED** [review]
@sanjaydasgupta Other than my minor comment ("should" -> "must"), I have nothing to add -- looks great.  Before we merge, let us just make sure that everything, including links render properly (`mkdocs`) -- and then please merge, or let me know, and I will in the morning Pacific Time.  Thanks a lot for this important documentation change!

File: docs/configuration/large_language_model.md [hunk 13]
```diff
@@ -88,20 +88,84 @@ ludwig train ...
...
-configuration options.
+The value of the `name` attribute (`input` in the example) is the name of a 
+dataset column. 
+
+See [Text Features](./features/text_features.md) for configuration options.
+
+### Single Dataset Column with Additional Text
+
+If the input to the LLM must be created by prefixing and/or suffixing some static text 
+to the content of one dataset column, then a `prompt` `template` should be provided to specify how
```
[the remark this answers is not in the document]
Comment (sanjaydasgupta) [reply docs/configuration/large_language_model.md:115]: I agree, will change and let you know.

**Comment by sanjaydasgupta** [comment]
I'm not sure how to complete the `mkdocs` part, but will read up. It will likely be late morning PT when I finish. Thanks for your help!

**Comment by alexsherstinsky** [comment]
> I'm not sure how to complete the `mkdocs` part, but will read up. It will likely be late morning PT when I finish. Thanks for your help!

No worries, @sanjaydasgupta -- there is a README file which describes how to launch `mkdocs` -- you basicaly pip install ludwig and then run mkdocs and connect to localhost on a port it says, then navigate to your pages and make sure that things look good to you.  Thank you!

**Comment by sanjaydasgupta** [comment]
> > I'm not sure how to complete the `mkdocs` part, but will read up. It will likely be late morning PT when I finish. Thanks for your help!
> 
> No worries, @sanjaydasgupta -- there is a README file which describes how to launch `mkdocs` -- you basicaly pip install ludwig and then run mkdocs and connect to localhost on a port it says, then navigate to your pages and make sure that things look good to you. Thank you!

Hi @alexsherstinsky, all complete now. I have checked using `mkdocs` too, and the text displayed has all the changes we made. The index in the right margin is also linking correctly to the three subheadings. Thanks for your help.
